src/adapters/angle-protocol/products/savings/angleProtocolSavingsAdapter.ts

```diff
@@ -0,0 +1,387 @@
+        underlyingTokenAddresses.set(key, await contract.asset())
+      }),
+    )
+
+    const protocolTokens: Map<Stablecoin, Erc20Metadata> = new Map()
+    await Promise.all(
+      Array.from(savingsContract).map(async ([key, contract]) => {
+        protocolTokens.set(
+          key,
+          await getTokenMetadata(
+            await contract.getAddress(),
+            this.chainId,
+            this.provider,
+          ),
+        )
+      }),
+    )
+
+    const underlyingTokens: Map<Stablecoin, Erc20Metadata> = new Map()
+    await Promise.all(
+      Array.from(underlyingTokenAddresses).map(async ([key, address]) => {
+        underlyingTokens.set(
+          key,
+          await getTokenMetadata(address, this.chainId, this.provider),
+        )
+      }),
+    )
+
+    const metadataObject: AngleProtocolMetadata = Array.from(
+      protocolTokens,
+    ).map(([key, protocolToken]) => ({
+      protocolToken,
+      underlyingToken: underlyingTokens.get(key)!,
+    }))
+
+    return metadataObject
+  }
+
+  async getProtocolTokens(): Promise<Erc20Metadata[]> {
+    const metadata = await this.buildMetadata()
+    return metadata.map((m) => m.protocolToken)
+  }
+
+  async getPositions(_input: GetPositionsInput): Promise<ProtocolPosition[]> {
+    const tokens = await this.buildMetadata()
+
+    const positions = await Promise.all(
+      tokens.map(async (token) => {
+        const savingContract = Savings__factory.connect(
+          token.protocolToken.address,
+          this.provider,
+        )
+        const balance = await savingContract.balanceOf(_input.userAddress, {
+          blockTag: _input.blockNumber,
+        })
+        const assetsBalance = await savingContract.convertToAssets(balance, {
+          blockTag: _input.blockNumber,
+        })
+        return {
+          ...token.protocolToken,
+          type: TokenType.Protocol,
+          balanceRaw: balance,
+          tokens: [
+            {
+              ...token.underlyingToken,
+              type: TokenType.Underlying,
+              balanceRaw: assetsBalance,
+            },
+          ],
+        }
+      }),
+    )
+
+    if (!_input.protocolTokenAddresses) return positions
+    return positions.filter((p) =>
+      _input.protocolTokenAddresses!.includes(p.address),
+    )
+  }
+
+  async getWithdrawals(_input: GetEventsInput): Promise<MovementsByBlock[]> {
+    const tokens = await this.buildMetadata()
+    const token = tokens.find(
+      (t) => t.protocolToken.address === _input.protocolTokenAddress,
+    )!
+
+    const savingContract = Savings__factory.connect(
+      _input.protocolTokenAddress,
+      this.provider,
+    )
+
+    const filter = savingContract.filters.Withdraw(
+      undefined,
+      _input.userAddress,
+    )
+    const eventResults = await savingContract.queryFilter(
+      filter,
+      _input.fromBlock,
+      _input.toBlock,
+    )
+
+    return eventResults.map((event) => ({
```

Comment:
don't need to provide the sahres removed too?

Reply:
If I am not mistaking the shares removed still emits a Withdraw event

Reply:
Yes but in the below object we only provide the `event.args?.assets` and not the burnt shares `event.args?.shares`